src/AzureBlobStorageUpdateSession/UpdateSession/AzureTableUpdateSessionManager.cs

```diff
@@ -0,0 +1,96 @@
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Etg.Yams.Update;
+
+namespace Etg.Yams.Azure.UpdateSession
+{
+    public class AzureTableUpdateSessionManager : IUpdateSessionManager
+    {
+        public const string UpdateSessionTableName = "YamsUpdateSession";
+        private readonly IUpdateSessionTable _updateSessionTable;
+        private readonly string _clusterId;
+        private readonly string _instanceId;
+        private readonly string _instanceUpdateDomain;
+
+        public AzureTableUpdateSessionManager(IUpdateSessionTable updateSessionTable, string clusterId, 
+            string instanceId, string instanceUpdateDomain)
+        {
+            _updateSessionTable = updateSessionTable;
+            _clusterId = clusterId;
+            _instanceId = instanceId;
+            _instanceUpdateDomain = instanceUpdateDomain;
+        }
+
+        public async Task<bool> TryStartUpdateSession(string appId)
+        {
+            Trace.TraceInformation(
+                $"Instance {_instanceId} will attempt to start update session for " +
+                $"ApplicationId = {appId}, UpdateDomain = {_instanceUpdateDomain}");
+
+            UpdateSessionTransaction transaction = new UpdateSessionTransaction(_clusterId, _instanceId, _instanceUpdateDomain, appId);
+            UpdateSessionStatus updateSessionStatus = await _updateSessionTable.FetchUpdateSessionStatus(_clusterId, appId);
+
+            if (updateSessionStatus.UpdateDomainEntity == null ||
+                updateSessionStatus.UpdateDomainEntity.UpdateDomain == _instanceUpdateDomain)
+            {
+                if (updateSessionStatus.UpdateDomainEntity == null)
+                {
+                    transaction.InsertUpdateDomain();
+                }
+                
+                transaction.MarkInstanceListAsModified();
+            }
+            else if(!updateSessionStatus.InstancesEntities.Any()) // no instance in the current update domain is updating
+            {
+                // set a new update domain (if no other instance beats us to it)
+                transaction.ReplaceUpdateDomain(updateSessionStatus); // will fail if current update domain changes
+                transaction.FailIfInstanceListModified(updateSessionStatus); // will fail if instance list changes
```

Comment:
Do this in ReplaceUpdateDomain? No other calls to FailIfInstanceListModified...

Reply:
I could but then I'd be giving the transaction class knowledge about how I am managing update domains, which is not part of its responsibility.